File: algorithms/206/main.go
package main

import (
	"bytes"
	"strconv"
	"fmt"
)

// Reverse a singly linked list.
// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func (l *ListNode) String() string {
	var buf bytes.Buffer
	buf.WriteByte('[')

	for l != nil {
		buf.WriteString(strconv.Itoa(l.Val))
		l = l.Next
		buf.WriteByte(' ')
	}
	buf.WriteByte(']')
	return buf.String()
}

// recursive
// Runtime: 3ms
func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	last, cur := head, head.Next
	head.Next = nil
	for cur != nil {
		next := cur.Next
		cur.Next = last
		last, cur = cur, next
	}
	return last
}

func main() {
	l := &ListNode{1, nil}
	fmt.Println(reverseList(l))

	l = &ListNode{1, &ListNode{2, nil}}
	fmt.Println(reverseList(l))

	l = &ListNode{1, &ListNode{2, &ListNode{3, nil}}}
	fmt.Println(reverseList(l))
}

File: algorithms/206/c++/main.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  explicit ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
  ListNode *reverseList(ListNode *head) {
    if (head == nullptr)
      return nullptr;
    ListNode *next = head->next, *nhead = head;
    if (next != nullptr) {
      nhead = reverseList(next);
      head->next = nullptr;
      next->next = head;
    }
    return nhead;
  }
};

int main(int argc, char *argv[]) {
  Solution so;
  ListNode n1{1};
  ListNode n2{2};
  n1.next = &n2;
  so.reverseList(&n1);
  return 0;
}